glommio/Cargo.toml

```diff
@@ -17,6 +17,7 @@ readme = "../README.md"
 log = "0.4"
 concurrent-queue = "1.1.2"
 futures-lite = "1.11.1"
+futures = "0.3.12"
```

Comment:
We have been trying to standardize on `futures-lite` and avoid using `futures`.
See 6405b59d0de6604d28ac7b857653b3b972a9c6cb for details

Reply:
Do you mind if I change the `connect` methods on `SharedSender`/`SharedReciever` a little bit like this, so that connections could be fired at once, and be waited one by one without potential deadlock, thus the dependency on `join` could be removed.

```rust
    pub fn connect(mut self) -> impl Future<Output = ConnectedSender<T>> {
        let state = self.state.take().unwrap();
        let reactor = Local::get_reactor();
        state.buffer.connect(reactor.id());
        let id = reactor.register_shared_channel(Box::new(enclose! {(state) move || {
            if state.buffer.consumer_disconnected() {
                state.buffer.capacity()
            } else {
                state.buffer.free_space()
            }
        }}));

        let reactor = Rc::downgrade(&reactor);
        let peer = Connector::new(state.buffer.clone(), reactor.clone());
        
        async move {
            let notifier = peer.await;
            ConnectedSender {
                state,
                id,
                reactor,
                notifier,
            }
        }
    }

```

Reply:
another alternative is to fire the connections `Task::local`, detach, and wait for the result. That is what I am doing in my N:M non-mesh implementation.

Reply:
`Task::local` requires it argument have a static lifetime, which is quite annoying. I tried removing it and glommio get compiled without problem. So why is it necessary for the future to be 'static?
```rust
    pub fn local(future: impl Future<Output = T> + 'static) -> Task<T>
    where
        T: 'static,
    {
        LOCAL_EX.with(|local_ex| local_ex.spawn(future))
    }
```

Reply:
Yes, good question. I wanted to discuss that too. For thread::spawn that is understandable. But in our case we copy future into the heap during the spawn, so we support static lifetime ourselves without requiring it to be static on the language level (we use reference counting). So IMHO it can be removed. @glommer WDYT ?

Reply:
it needs to be static because you can `.detach` it. If you don't `.detach` it you can probably figure out a lifetime for it but at task creation you don't know if it will be detached or not.

You can experiment with that for yourself: it is easy to get into a situation where you remove the `'static` requirement, which allows you to now pass an `async {}` closure that uses a reference from the outside. Add a timer to the task and access that reference after the timer. Outside the `Task`, the reference goes out of scope and gets freed -> boom.

Reply:
Simplified a little bit to remove dependency on self, now I'm able to use `Task::local`.

Reply:
But that is possible to lower restrictions any way by providing a version of the method which prohibits detaching of the task.
For example 
```
pub async fn spawn(future: impl Future<Output = T>) 
```
such a task can not be detached and as result does not require a static lifetime. Borrow checker is a pain point for many newcomers so lowering restrictions worth to consider. It is helpful even for more experienced developers. @glommer WDYT ?

BTW local IMHO is not the best name for the method. Would be nice to have `spawn_detachable` and `spawn` instead.

Reply:
If join handle dies and the task is not stored in weaker holders (I am considering the case that user doesn't call await) then the task is dropped on Drop so it will not outlive internal data. So such an approach seems safe for me.

Reply:
I am trying to understand what exactly it is that you are proposing, @laa . Is the idea that you would be able to spawn a task that would start running in the background?

If we can do:
```
let a = spawn(); // tasks starts running
... more async code here

a.await; // waits for a to finish, might have finished already
```

I see value on that.

About `local`, I agree. I took it from `async-task`, that had `spawn` and `local` for a local spawn. However for us, everything is local and it is a bad name.

We're being very lax prior to 1.0 with the API, but `local` is a bit different because it is used a lot by everyone. I would be okay to remove that only after deprecation.

Reply:
Hi @glommer , let us continue later in community chat. Now I am quite busy with allocator and can not spend enough time to completely analyze the idea.  As for the current moment, I incline to think that better to leave all as is. I need to think more about that topic :-) .